File: src/main/java/me/itzisonn_/meazy/parser/ast/statement/ForeachStatement.java
package me.itzisonn_.meazy.parser.ast.statement;

import kotlin.Unit;
import lombok.Getter;
import me.itzisonn_.meazy.parser.ast.ProgramUnit;
import me.itzisonn_.meazy.registry.Registries;
import me.itzisonn_.meazy.instruction.InstructionsSet;
import me.itzisonn_.meazy.instruction.method.InvokeMethodInstruction.InvokeType;
import me.itzisonn_.meazy.parser.DataType;
import me.itzisonn_.meazy.parser.ast.expression.Expression;
import me.itzisonn_.meazy.runtime.environment.Environment;
import me.itzisonn_.meazy.runtime.environment.LocalVariableDeclarationEnvironment;
import me.itzisonn_.meazy.runtime.environment.LoopEnvironment;
import me.itzisonn_.meazy.runtime.VariableValue;
import org.jspecify.annotations.NullMarked;

import java.lang.constant.ClassDesc;
import java.lang.constant.ConstantDescs;
import java.lang.constant.MethodTypeDesc;
import java.util.List;

@Getter
@NullMarked
public class ForeachStatement implements LocalStatement {
    private final boolean isConstant;
    private final String id;
    private final DataType dataType;
    private final Expression collection;
    private final List<LocalStatement> body;

    public ForeachStatement(boolean isConstant, String id, DataType dataType, Expression collection, List<LocalStatement> body) {
        this.isConstant = isConstant;
        this.id = id;
        this.dataType = dataType;
        this.collection = collection;
        this.body = body;
    }

    @Override
    public void emit(InstructionsSet instructionsSet, Environment environment, ProgramUnit parent) {
        if (!(environment instanceof LocalVariableDeclarationEnvironment localVariableDeclarationEnvironment)) {
            throw new RuntimeException("Foreach statement must be inside variableDeclarationEnvironment TODO");
        }

        VariableValue iterableVariableValue = localVariableDeclarationEnvironment.declareVariable(
                null,
                DataType.ofNonNull(ClassDesc.of("java.lang.Iterable")),
                true,
                null
        );

        collection.emit(instructionsSet, environment, this);

        instructionsSet.invokeMethod(
                ClassDesc.of("java.lang.Iterable"),
                "iterator", 
                MethodTypeDesc.of(ClassDesc.of("java.util.Iterator")),
                _ -> Unit.INSTANCE,
                InvokeType.INTERFACE
        );

        instructionsSet.storeLocal(ClassDesc.of("java.util.Iterator"), iterableVariableValue.getSlot());

        var conditionLabel = instructionsSet.createAndInitLabel();
        var endLabel = instructionsSet.createAndInitLabel();
        LoopEnvironment loopEnvironment = Registries.LOOP_ENVIRONMENT_FACTORY.getEntry().getValue().create(environment, conditionLabel, endLabel);

        instructionsSet.bindLabel(conditionLabel);

        instructionsSet.getLocal(ClassDesc.of("java.util.Iterator"), iterableVariableValue.getSlot());
        instructionsSet.invokeMethod(
                ClassDesc.of("java.util.Iterator"),
                "hasNext",
                MethodTypeDesc.of(ConstantDescs.CD_boolean),
                _ -> Unit.INSTANCE,
                InvokeType.INTERFACE
        );
        instructionsSet.gotoLabelIfEqualsZero(endLabel);

        instructionsSet.getLocal(ClassDesc.of("java.util.Iterator"), iterableVariableValue.getSlot());
        instructionsSet.invokeMethod(
                ClassDesc.of("java.util.Iterator"),
                "next",
                MethodTypeDesc.of(ConstantDescs.CD_Object),
                _ -> Unit.INSTANCE,
                InvokeType.INTERFACE
        );

        dataType.resolve(environment);
        instructionsSet.checkCast(dataType.getClassDesc());

        VariableValue variableValue = localVariableDeclarationEnvironment.declareVariable(id, dataType, isConstant, null);
        instructionsSet.storeLocal(variableValue.getDataType().getClassDesc(), variableValue.getSlot());
        instructionsSet.setLocalName(variableValue.getSlot(), variableValue.getId(), variableValue.getDataType().getClassDesc(), conditionLabel, endLabel);

        for (Statement statement : body) {
            statement.emit(instructionsSet, loopEnvironment, this);
        }

        instructionsSet.gotoLabel(conditionLabel);
        instructionsSet.bindLabel(endLabel);
    }

    @Override
    public boolean alwaysReturns() {
        for (LocalStatement localStatement : body) {
            if (localStatement.alwaysReturns()) return true;
        }

        return false;
    }
}

File: src/main/kotlin/me/itzisonn_/meazy/parser/ast/statement/ForeachStatement.kt
package me.itzisonn_.meazy.parser.ast.statement

import me.itzisonn_.meazy.instruction.InstructionsSet
import me.itzisonn_.meazy.instruction.method.InvokeMethodInstruction.InvokeType
import me.itzisonn_.meazy.parser.DataType
import me.itzisonn_.meazy.parser.ast.ProgramUnit
import me.itzisonn_.meazy.parser.ast.expression.Expression
import me.itzisonn_.meazy.runtime.environment.Environment
import me.itzisonn_.meazy.runtime.environment.LoopEnvironment
import me.itzisonn_.meazy.runtime.environment.declaration.LocalVariableDeclarationEnvironment
import java.lang.constant.ClassDesc
import java.lang.constant.ConstantDescs
import java.lang.constant.MethodTypeDesc

class ForeachStatement(
    val isConstant: Boolean,
    val id: String,
    val dataType: DataType,
    val collection: Expression,
    val body: List<LocalStatement>
) : LocalStatement {
    override fun emit(instructions: InstructionsSet, environment: Environment, parent: ProgramUnit) {
        if (environment !is LocalVariableDeclarationEnvironment) {
            throw RuntimeException("Foreach statement must be inside variableDeclarationEnvironment TODO")
        }

        val iterableVariableValue = environment.declareVariable(
            DataType.ofNonNull(ClassDesc.of("java.lang.Iterable")),
            true,
            null
        )

        collection.emit(instructions, environment, this)

        instructions.invokeMethod(
            ClassDesc.of("java.lang.Iterable"),
            "iterator",
            MethodTypeDesc.of(ClassDesc.of("java.util.Iterator")),
            InvokeType.INTERFACE
        )

        instructions.storeLocal(ClassDesc.of("java.util.Iterator"), iterableVariableValue.slot)

        val conditionLabel = instructions.createAndInitLabel()
        val endLabel = instructions.createAndInitLabel()
        val loopEnvironment = LoopEnvironment(environment, conditionLabel, endLabel)

        instructions.bindLabel(conditionLabel)

        instructions.getLocal(ClassDesc.of("java.util.Iterator"), iterableVariableValue.slot)
        instructions.invokeMethod(
            ClassDesc.of("java.util.Iterator"),
            "hasNext",
            MethodTypeDesc.of(ConstantDescs.CD_boolean),
            InvokeType.INTERFACE
        )
        instructions.gotoLabelIfEqualsZero(endLabel)

        instructions.getLocal(ClassDesc.of("java.util.Iterator"), iterableVariableValue.slot)
        instructions.invokeMethod(
            ClassDesc.of("java.util.Iterator"),
            "next",
            MethodTypeDesc.of(ConstantDescs.CD_Object),
            InvokeType.INTERFACE
        )

        dataType.resolve(environment)
        instructions.checkCast(dataType.classDesc)

        val variableValue = environment.declareVariable(id, dataType, isConstant, null)
        instructions.storeLocal(variableValue.dataType.classDesc, variableValue.slot)
        instructions.setLocalName(
            variableValue.slot,
            id,
            variableValue.dataType.classDesc,
            conditionLabel,
            endLabel
        )

        for (statement in body) {
            statement.emit(instructions, loopEnvironment, this)
        }

        instructions.gotoLabel(conditionLabel)
        instructions.bindLabel(endLabel)
    }

    override fun alwaysReturns(): Boolean {
        for (localStatement in body) {
            if (localStatement.alwaysReturns()) return true
        }

        return false
    }
}
